tooling/nargo_cli/src/cli/fv_cmd.rs

```diff
@@ -58,16 +65,254 @@ pub(crate) fn run(args: FormalVerifyCommand, workspace: Workspace) -> Result<(),
+    let lines: Vec<String> = stderr_output.lines().map(String::from).collect();
+    for line in &lines {
+        let smt_output = serde_json::from_str::<SmtOutput>(&line).map_err(|e| {
+            CliError::Generic(format!(
+                "Failed to deserialize the Venir output: {} with the following error\n{}",
+                line,
+                e.to_string()
+            ))
+        })?;
+
+        smt_outputs.push(smt_output);
+    }
+    smt_outputs.reverse();
+
+    let mut verification_diagnostics: Vec<CustomDiagnostic> = smt_outputs
+        .into_iter()
+        .map(|smt_output| smt_output_to_diagnostic(smt_output, &workspace_file_manager))
+        .collect::<Result<_, _>>()?;
+
+    // Sort errors by span.
+    verification_diagnostics.sort_by(|a, b| {
+        match (a.secondaries.first(), b.secondaries.first()) {
+            (None, None) => std::cmp::Ordering::Equal, // Errors with no span are put at the start.
+            (None, Some(_)) => std::cmp::Ordering::Less,
+            (Some(_), None) => std::cmp::Ordering::Greater,
+            (Some(a_custom_label), Some(b_custom_label)) => {
+                a_custom_label.location.span.start().cmp(&b_custom_label.location.span.start())
+            }
+        }
+    });
+
+    // Report errors from the verification process.
+    let reported_errors: ReportedErrors = noirc_errors::reporter::report_all(
+        workspace_file_manager.as_file_map(),
+        &verification_diagnostics,
+        deny_warnings,
+        false,
+    );
+
+    if reported_errors.error_count == 0 {
+        println!("Verification successful!");
+        Ok(())
+    } else {
+        if has_crashed {
+            Err(CliError::Generic(format!("Verification crashed!")))
+        } else {
+            Err(CliError::Generic(format!(
+                "Verification failed due to {} previous errors!",
+                reported_errors.error_count
+            )))
+        }
+    }
+}
+
+/// Part of the Venir output standard.
+#[derive(Deserialize)]
+struct ErrorBlock {
+    error_message: String,
+    error_span: String,
+    secondary_message: String,
+}
+
+/// Part of the Venir output standard.
+#[derive(Deserialize)]
+struct WarningBlock {
+    warning_message: String,
+}
+
+/// Part of the Venir output standard.
+#[derive(Deserialize)]
+struct CrashBlock {
+    crash_message: String,
+    crash_span: String,
+}
+
+/// The possible outputs of the Venir binary.
+#[derive(Deserialize)]
+enum SmtOutput {
+    Error(ErrorBlock),
+    Warning(WarningBlock),
+    Note(String),
+    AirMessage(CrashBlock),
+}
+
+/// Maps a Venir output to a Noir diagnostic type error.
+fn smt_output_to_diagnostic(
+    smt_output: SmtOutput,
+    workspace_file_manager: &FileManager,
+) -> Result<CustomDiagnostic, CliError> {
+    let default_file_id = workspace_file_manager
+        .as_file_map()
+        .all_file_ids()
+        .last()
+        .unwrap_or(&FileId::dummy())
+        .clone();
```

Comment:
Prefer `.cloned().unwrap_or(Field::dummy())` to skin an unnecessary `.clone()` of the default value.